Create a new split container when switching a workspace container to split layout

9ary opened the pull request:
The behavior before 52ce8c8 was to do it regardless of what layout we're
switching to.

Fixes #2846

Like I said in the issue thread, I need to check things out a bit more if tests are wanted, but I'm leaving this here in the meantime.

9ary committed:
Create a new split container when switching a workspace container to split layout

The behavior before 52ce8c8 was to do it regardless of what layout we're
switching to.

Fixes #2846

acrisci commented:
Yeah we definitely need a regression test for this.

Streetwalrus commented:
I'm unable to get the test suite running.
```cpanm (App::cpanminus) 1.7043 on perl 5.026000 built for x86_64-linux-thread-multi
Work directory is /home/streetwalrus/.cpanm/work/1501446269.21773
You have make /usr/bin/make
You have LWP 6.26
You have /usr/bin/tar: tar (GNU tar) 1.29
Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
You have /usr/bin/unzip
--> Working on .
Entering /home/streetwalrus/i3/testcases
Configuring /home/streetwalrus/i3/testcases
Running Makefile.PL
Warning: NAME must be a package name
Generating a Unix-style Makefile
Writing Makefile for i3-testsuite
Writing MYMETA.yml and MYMETA.json
-> OK
Checking dependencies from MYMETA.json ...
Checking if you have IPC::Run 0 ... Yes (0.96)
Checking if you have Inline 0 ... Yes (0.80)
Checking if you have AnyEvent 0 ... Yes (7.14)
Checking if you have Test::More 0.94 ... Yes (1.302073)
Checking if you have AnyEvent::I3 0.16 ... Yes (0.17)
Checking if you have ExtUtils::MakeMaker 0 ... Yes (7.24)
Checking if you have ExtUtils::PkgConfig 0 ... Yes (1.16)
Checking if you have X11::XCB 0.12 ... Yes (0.16)
Checking if you have Inline::C 0 ... Yes (0.78)
Building and testing i3-testsuite-
Running Mkbootstrap for testsuite ()
chmod 644 "testsuite.bs"
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- testsuite.bs blib/arch/auto/i3-testsuite/testsuite.bs 644
rm -f blib/arch/auto/i3-testsuite/testsuite.so
cc  -shared -Wl,-O1,--sort-common,--as-needed,-z,relro -L/usr/local/lib -fstack-protector-strong testsuite.o  -o blib/arch/auto/i3-testsuite/testsuite.so  \
      \
  
cc: error: testsuite.o: No such file or directory
make: *** [Makefile:470: blib/arch/auto/i3-testsuite/testsuite.so] Error 1
-> FAIL Installing . failed. See /home/streetwalrus/.cpanm/work/1501446269.21773/build.log for details. Retry with --force to force install it.```

Any idea?

hwangcc23 committed:
Add regression tests for #2846

1). Add one regression test in 167-workspace_layout.t:
    - Get a fresh workspace
    - Set the layout to something
    - Create windows
    - Try to switch to another layout
    - Check if successful
    - Repeat for all 12 possible transitions
2). Add another regression test in 167-workspace_layout.t:
    - Check that the command 'layout toggle split' works regardless of
      what layout we're using

hwangcc23 commented:
@Streetwalrus I don't see this error before. What is the command you used to activate the testsuite?
P.S. I have written a test case for the regression test. Feel free to reference it or use it directly.
[https://github.com/hwangcc23/i3/commit/f234bf3428a16b4c7b94c5659ed839a8e01f796e](url)

Streetwalrus commented:
I ran `cpanm .` as instructed in the documentation. The issue could be that I was trying to run it as a regular user to avoid installing things to my system that aren't managed by pacman.

Thanks for writing that test. I think it would be better this way though:
- Get a fresh workspace
- Set the layout to something
- Create one window (or two)
- Try to switch to another layout
- Check if successful
- Repeat for either 4 transitions (with 4 distinct targets) or all 12 possible transitions

I don't think `layout toggle split` really matters here, considering the code in question doesn't care about toggling, only about setting a new layout.

hwangcc23 commented:
@Streetwalrus Do you still work on this?
I have written some test cases for testing all transitions. Feel free to reference them, or ask me to send the PR.
https://github.com/hwangcc23/i3/commit/49f2d81d746872d8428263cc5d21f5427749c7c0

Streetwalrus commented:
Thanks a lot. I'll cherry-pick them into the pr if that's ok with you.

hwangcc23 commented:
@Streetwalrus OK.  :)

Streetwalrus commented:
@hwangcc23, I've just looked at your tests, in order to reproduce the bug, you need to set the initial layout before spawning the first window, then change it afterwards. You don't really need two windows to check that the layout was successfully changed either.

hwangcc23 commented:
https://github.com/hwangcc23/i3/commit/c7437c856d4c9a36fddfb0e316ba14b2a8c6864b

Streetwalrus commented:
Bump. Not sure why Travis failed as the tests seem fine in the log.

Airblader commented:
@Streetwalrus I'm a bit lost – what is the status here? Can it be reviewed from your perspective? If so, please rebase it so that the rests run again; we fixed a lot over the last few days regarding the travis setup.

Streetwalrus commented:
As far as I'm concerned the patch is fine, I've been running it ever since I've opened this PR (frequently merging master). I'll rebase it right away.

Airblader commented on testcases/t/167-workspace_layout.t:
```diff
@@ -375,7 +375,209 @@ ok(@content == 2, 'two containers opened');
+
+#####################################################################
+# 16: Check that the command 'layout toggle split' works regardless
+# of what layout we're
```
Something is missing here.

Airblader commented on testcases/t/167-workspace_layout.t:
```diff
@@ -375,7 +375,209 @@ ok(@content == 2, 'two containers opened');
+workspace_layout default
+EOT
+
+$pid = launch_with_config($config);
```
Could you please adapt the tests to use `i3_config` instead of `launch_with_config` / `exit_gracefully`, see https://github.com/i3/i3/pull/2945/commits/02786155dc7f5962242681f53383a54230f76965?

Streetwalrus commented:
@hwangcc23 since you wrote the tests, would you like to look into that?

hwangcc23 commented:
@Streetwalrus It is fixed. Please cherry-pick them again.
https://github.com/hwangcc23/i3/commit/b5a0bc1370231fc719c5de80e193a5709ad5c765

Streetwalrus commented:
Pulled in the new commit and rebased. @Airblader, you requested something else than the typo but apparently deleted the comment?

Streetwalrus commented:
The test is currently failing because when the initial layout of a workspace is not a split, [this](https://github.com/i3/i3/blob/9fe508b79163e74ea61d4abf432f06eea216cb6a/src/con.c#L1881-L1883) skips changing the layout. Removing this check seems to make another test fail.
The `toggle split` test also seems to assume that the screen is in the landscape orientation, a portrait screen would have the default split be vertical, which would make it fail.

9ary committed:
Default to L_SPLITH with toggle split when last_split_layout hasn't been initialized

Streetwalrus commented:
I ended up following the behavior in `con_set_layout`, which is simply to default to a horizontal split.

Airblader commented:
All those test cases are the same with a few different values plugged in. Can we refactor them to iterate over those values and have the actual implementation of the test only once?

domenkozar commented:
FYI: `layout splitv` still works.

Streetwalrus commented:
>Can we refactor them to iterate over those values and have the actual implementation of the test only once?

I'll check it out. Guess it's time to learn some Perl.

Streetwalrus commented:
This should do the trick.

Airblader reviewed (approved):
Thanks!

cpick commented:
For anyone running Ubuntu Bionic, I created a PPA that applies this patch to the 4.14.1-1 version of `i3` that bionic ships with:

    sudo add-apt-repository ppa:cpick/i3
    sudo apt update
    sudo apt install i3
